orttraining/orttraining/test/graph/training_session_test_utils.h

```diff
@@ -0,0 +1,525 @@
+    Tensor actual_tensor{actual_gpu_tensor.DataType(), actual_gpu_tensor.Shape(), cpu_allocator};
+    ORT_ENFORCE(data_transfer_mgr.CopyTensor(actual_gpu_tensor, actual_tensor).IsOK());
+#else
+    auto& actual_tensor = a_state_it.second.Get<Tensor>();
+#endif
+    if (expected_tensor.GetElementType() == ONNX_NAMESPACE::TensorProto_DataType_INT64) {
+      // compare "Update_Count" or "Step"
+      ASSERT_EQ(actual_tensor.GetElementType(), ONNX_NAMESPACE::TensorProto_DataType_INT64);
+      ASSERT_EQ(expected_tensor.Shape(), actual_tensor.Shape());
+      std::vector<int64_t> dims = {1};
+      ASSERT_EQ(expected_tensor.Shape().GetDims(), dims);
+      auto size = expected_tensor.Shape().Size();
+      const std::vector<int64_t> expected(expected_tensor.template Data<int64_t>(), expected_tensor.template Data<int64_t>() + size);
+      const std::vector<int64_t> actual(actual_tensor.template Data<int64_t>(), actual_tensor.template Data<int64_t>() + size);
+      // the step value will be incremented by 1 after a train step
+      ASSERT_EQ(expected[0] + 1, actual[0]);
+    } else if (expected_tensor.GetElementType() == ONNX_NAMESPACE::TensorProto_DataType_FLOAT ||
+               expected_tensor.GetElementType() == ONNX_NAMESPACE::TensorProto_DataType_FLOAT16) {
+      // adding a tolerance as after a train step, the moment tensor value will be updated
+      horizontal_parallel_test_utils::VerifyOutputs(expected_tensor, actual_tensor, true, 1e-8f, 1e-7f, 0.32f);
+    }
+  }
+}
+
+void VerifyOptimizerState(const DataTransferManager& data_transfer_manager, const TrainingSession::OptimizerState& expected_state, const TrainingSession::OptimizerState& actual_state) {
+  for (const auto& a_state_it : actual_state) {
+    std::string key = a_state_it.first;
+    const auto& e_state_it = expected_state.find(key);
+    ORT_ENFORCE(e_state_it != expected_state.end());
+    VerifyState(data_transfer_manager, e_state_it->second, a_state_it.second);
+  }
+}
+
+std::unordered_set<std::string> GetModelOutputNames(const InferenceSession& session) {
+  const auto outputs_result = session.GetModelOutputs();
+  ORT_ENFORCE(outputs_result.first.IsOK(), "Failed to get model outputs: ", outputs_result.first.ErrorMessage());
+  std::unordered_set<std::string> output_names{};
+  for (const auto* output : *outputs_result.second) {
+    output_names.insert(output->Name());
+  }
+  return output_names;
+}
+}  // namespace
+
+static TrainingSession::TrainingConfiguration MakeBasicTrainingConfig() {
+  TrainingSession::TrainingConfiguration config{};
+  config.model_with_training_graph_path = BACKWARD_MODEL_PATH;
+  config.loss_function_config = TrainingSession::TrainingConfiguration::LossFunctionConfiguration{};
+  config.loss_function_config.value().loss_function_info =
+      LossFunctionInfo(OpDef("MeanSquaredError"), "loss", {"predictions", "labels"});
+
+  return config;
+}
+
+static Status BuildBackPropGraph(
```

Comment:
This was not used

Reply:
it is used in another file, but moved it out